sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -249,10 +264,8 @@ def _random_sample_mask(cnp.npy_intp n_total_samples,
          An ndarray where ``n_total_in_bag`` elements are set to ``True``
          the others are ``False``.
      """
-     cdef cnp.ndarray[float64, ndim=1, mode="c"] rand = \
-          random_state.uniform(size=n_total_samples)
-     cdef cnp.ndarray[uint8, ndim=1, mode="c", cast=True] sample_mask = \
-          np_zeros((n_total_samples,), dtype=bool)
+     cdef cnp.float64_t[::1] rand = random_state.uniform(size=n_total_samples)
+     cdef cnp.uint8_t[::1] sample_mask = np_zeros((n_total_samples,), dtype=bool)
```

Comment:
I agree that might possibly be the reason.

Reply:
The following should be better then.

```suggestion
     cdef cnp.uint8_t[::1] sample_mask = np_zeros((n_total_samples,), dtype=np.uint8)
```

Reply:
I tried this but it does not seem to work. That is because we might actually need the underlying numpy array to be a boolean one. The function is used at this point https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/ensemble/_gb.py#L603 and the returned array is directly used to index into other arrays:
 y, raw_predictions and sample_weight.